### PR TITLE
Implement oms_rename for sub-models

### PR DESCRIPTION
**Pull request by lochel**
### Related Issues

* #926.
* OpenModelica/OpenModelica#7118

### Purpose

The `oms_rename` function does only work for model identifiers, but not for systems and components.

### Expected Behavior

```
oms_rename("modelA", "modelB")     // renaming a model from modelA to modelB
oms_rename("modelA.Root", "root")  // renaming system Root to root in modelA
oms_rename("modelA.Root.foo", "some_name")    // renaming an FMU in the Root system from foo to some_name 
```

### CODE REVIEW

File: testsuite/OMSimulator/rename.lua [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+oms_addSubModel("model.root.add", "../resources/Modelica.Blocks.Math.Add.fmu")
+
+src = oms_list("model")
+print(src)
+
+status = oms_rename("model.root", "root_1")
+status = oms_rename("model.root.System1", "System_1")
+status = oms_rename("model.root.add", "add_1")
+
```
Comment (lochel) [inline testsuite/OMSimulator/rename.lua:24]: I am a bit confused. Shouldn't it be `oms_rename("model.root_1.System1", "System_1")` and `oms_rename("model.root_1.add", "add_1")`?

Comment (arun3688) [reply testsuite/OMSimulator/rename.lua]: @lochel this should be fixed with this https://github.com/OpenModelica/OMSimulator/pull/928/commits/ad744631b2fc5b3e8cb13a324e98d5c0b0e7ccb5

**Comment by lochel** [comment]
@adeas31 this will cause OMEdit (with OpenModelica/OpenModelica#7118) to crash. Can you have a quick look?

File: src/OMSimulatorLib/Model.cpp [hunk 2]
```diff
@@ -136,17 +146,25 @@ oms_status_enu_t oms::Model::rename(const ComRef& cref, const ComRef& newCref)
...
   // rename submodules
   auto component = system->getComponents().find(tail);
   if (component != system->getComponents().end())
   {
-    return component->second->rename(newCref);
+    component->second->rename(newCref);
+    system->getComponents()[newCref] = component->second;
+    // delete old cref
+    system->getComponents().erase(component);
+    return oms_status_ok;
```
Comment (lochel) [inline src/OMSimulatorLib/Model.cpp]: try
```
component->first = newCref;
component->second->rename(newCref);
```

Comment (lochel) [reply src/OMSimulatorLib/Model.cpp]: never mind...

**Review by lochel — COMMENTED** [review]
(_OMSimulator v2.1.1.post6-g620bad6-linux_)
@arun3688 Renaming model A to B should fail if there exists already a model B.

File: src/OMSimulatorLib/Connection.cpp [hunk 3]
```diff
@@ -231,3 +231,46 @@ bool oms::Connection::isValid(const ComRef& crefA, const ComRef& crefB, const Co
...
+
+oms_status_enu_t oms::Connection::rename(const ComRef& cref, const ComRef& newCref)
+{
+  if (getSignalA() == cref)
+  {
+    std::string str(newCref);
+    delete [] this->conA;
+    this->conA = new char[str.size()+1];
+    strcpy(this->conA, str.c_str());
+  }
```
Comment (lochel) [inline src/OMSimulatorLib/Connection.cpp:243]: Remove this!

File: src/OMSimulatorLib/Connection.cpp [hunk 4]
```diff
@@ -231,3 +231,46 @@ bool oms::Connection::isValid(const ComRef& crefA, const ComRef& crefB, const Co
...
+
+  if (getSignalB() == cref)
+  {
+    std::string str(newCref);
+    delete [] this->conB;
+    this->conB = new char[str.size()+1];
+    strcpy(this->conB, str.c_str());
+  }
```
Comment (lochel) [inline src/OMSimulatorLib/Connection.cpp:262]: Remove this!

**Pull request merged**